### DISCUSSION TITLE
Come risalire ad una Proforma data una Fattura?

### DISCUSSION THREAD

**Discussion by nicolo-faedi** [fattureincloud/api#81] [Q&A] [answered]
Buongiorno,
Sto cercando di risalire e mettere in relazione le fatture e le relative proforma. 
Attualmente interrogando correttamente l'API vedo che nell'oggetto di type 'invoice' è presente un attributo chiamato 'original_document' che contiene l'attributo 'number' e 'date' relativi alla fattura proforma di riferimento.
Purtropo noto che molte date antecedenti il 2022 non corrispondono a quelle presenti nel database delle proforma, e quindi programmaticamente non posso metterle in relazione.
È un bug noto? Oppure sto sbagliando qualcosa?

Vi ringrazio per la collaborazione

**Comment by emajo** [comment, marked as answer]
Ciao @nicuzz, il modo migliore per cercare la proforma di riferimento sarebbe tramite `numeration + number` in quanto sicuramente univoco, tuttavia la data riportata in original document dovrebbe essere la medesima riportata effettivamente nella proforma.
Il problema riscontrato è un mismatch tra la data delle proforma salvate in locale e quelle salvate su FattureInCloud oppure tra la data delle proforma salvate in locale e le date riportate in original_document?

**Reply by nicolo-faedi** [reply]
Buon pomeriggio grazie per la risposta
Al momento il mismatch lo riscontro dai soli dati restituiti dall'API. 
Per step:

- Ottengo le invoice dall'endpoint `/c/:company_id/issued_documents?type=invoice&fields=original_document,date,number,numeration,id&per_page=50&sort=-date`
- Estraggo come chiave `number` + `date` da `original_document` delle invoice ricevute
- Ottengo le proforma dall'endpoint `/c/:company_id/issued_documents?type=proforma&per_page=50&sort=-date`
- Per qualche invoice trovo la proforma corrispondente number+date, per altre o non è presente, o il parametro numeration è "", oppure le date non corrispondono.
- Segnalo infine un altro malfunzionamento: quando interrogo l'endpoint `issued_documents` specificando `fields`, talvolta vengono restituiti **null** dei campi che invece con la query generica senza fields, sono impostati!


1. Le invoice (quelle dal 2022 in poi) mi ritornano correttamente la proforma di riferimento nell'original_document.
2. In quasi tutte le `invoice` precedenti al 2022 l'attributo `original_document` è **null** e quindi impossibile rintracciare la proforma di riferimento.
3. L'attributo `numeration` non è definito in tutti gli original document e quindi mi è impossibile fare l'associazione che mi avete proposto

Resto a disposizione
Cordialmente

**Reply by emajo** [reply]
Ciao @nicuzz, per quanto riguarda la ricerca per `numeration + number` dovrebbe essere fattibile anche quando numeration è vuoto in quanto number è univoco per qualsiasi numerazione, anche `""`, per quanto riguarda invece `specificando fields vengono restituiti null dei campi che invece con la query generica senza fields sono impostati` saremmo interessati a approfondire questo problema, se riesce a contattare il nostro supporto clienti fornendo informazioni per aiutarci a replicare sarebbe molto d'aiuto, stessa cosa per quanto riguarda i documenti per cui le date non corrispondono, fornendo gli id dei documenti al nostro supporto saremo in grado di capire cosa sta succedendo.

Per le invoice precedenti al 2022 `original_document` è null perchè la trasformazione è una feature che è stata aggiunta quest'anno, precedentemente i documenti venivano traformati creandone una copia e modificandola, perdendo così il legame con il documento originale.
